libs/langchain/langchain/automaton/runnables.py

```diff
@@ -0,0 +1,108 @@
+"""Module contains useful runnables for agents."""
+from __future__ import annotations
+
+from typing import Sequence, Callable, List, Optional, Union
+
+from langchain.automaton.typedefs import (
+    MessageLike,
+    MessageLog,
+    FunctionResult,
+    FunctionCall,
+)
+from langchain.schema import BaseMessage, AIMessage, PromptValue
+from langchain.schema.language_model import BaseLanguageModel
+from langchain.schema.runnable.base import RunnableLambda, Runnable
+from langchain.tools import BaseTool
+
+
+# PUBLIC API
+
+
+def create_tool_invoker(
+    tools: Sequence[BaseTool],
+) -> Runnable[FunctionCall, FunctionResult]:
+    """Re-write with router."""
+    tools_by_name = {tool.name: tool for tool in tools}
+
+    def func(function_call: FunctionCall) -> FunctionResult:
+        """A function that can invoke a tool using .run"""
+        try:
+            tool = tools_by_name[function_call.name]
+        except KeyError:
+            raise AssertionError(f"No such tool: {function_call.name}")
+        try:
+            result = tool.run(function_call.arguments or {})
+            error = None
+        except Exception as e:
+            result = None
+            error = repr(e) + repr(function_call.arguments)
+        return FunctionResult(name=function_call.name, result=result, error=error)
+
+    return RunnableLambda(func=func)
+
+
+LogOrMessages = Union[MessageLog, Sequence[MessageLike]]
+
+
+def create_llm_program(
+    llm: BaseLanguageModel,
+    prompt_generator: Callable[
+        [LogOrMessages], Union[str, PromptValue, Sequence[BaseMessage]]
+    ],
+    *,
+    tools: Optional[Sequence[BaseTool]] = None,
+    stop: Optional[Sequence[str]] = None,
+    parser: Union[
+        Runnable[Union[BaseMessage, str], MessageLike],
+        Callable[[Union[BaseMessage, str]], MessageLike],
+        None,
+    ] = None,
+    invoke_tools: bool = True,
+) -> Runnable[MessageLog, List[MessageLike]]:
+    """Create a runnable that can update memory."""
+
+    tool_invoker = create_tool_invoker(tools) if invoke_tools else None
+
+    def _bound(message_log: MessageLog) -> List[MessageLike]:
```

Comment:
IMO this should be written as a RunnableSequence, as that will 
1. Preserve tracing info 
2. Support batch/stream correctly if the steps support it, etc